### PR TITLE
Add dygraph double grad implementation

### PR DESCRIPTION
**Pull request by sneaxiy**
This PR adds double grad implementation in dygraph mode. The implementation is based on `paddle.fluid.dygraph.grad` which would calculate the gradients of `y` with respect to `x` (`x` and `y` can be any vars in the network). 

StarGAN models with gradient penalty are tested in this PR.

This PR also rewrites some original codes of dygraph, such as `VariableWrapper`, `SavedVariableWrapperList`, `GradOpNode`, etc. After testing in PTB model, the time cost of one epoch on V100 CUDA 9 machine is about 82-84s after this revision, which is the same as the original develop code.

### CODE REVIEW

File: paddle/fluid/imperative/partial_grad_engine.cc [hunk 1]
```diff
@@ -0,0 +1,1026 @@
...
+        }
+      }
+    }
+  }
+
+  // Prepare new outputs
+  NameVarMap<VarBase> tmp_outs;
+  std::unordered_map<GradientAccumulationInfo *,
+                     std::vector<std::shared_ptr<VariableWrapper>>>
+      grads_to_accumulate;
```
Comment (phlrain) [inline paddle/fluid/imperative/partial_grad_engine.cc]: 这个grads_to_accumulate 是不是不一定要用一个unorder_map, 后面也没有强的查询需求，可能会略微影响速度

File: paddle/fluid/imperative/partial_grad_engine.cc [hunk 2]
```diff
@@ -0,0 +1,1026 @@
...
+    }
+  }
+
+  VLOG(10) << "Created " << double_grad_nodes_.size() << " double grad ops";
+  return CreateResult();
+}
+
+void PartialGradTask::RunEachOp(const OpBase *op) {
+  // Prepare new inputs
+  NameVarMap<VarBase> tmp_ins;
```
Comment (phlrain) [inline paddle/fluid/imperative/partial_grad_engine.cc:778]: 如果 create_graph = False 情况下 是不是没有必要创建tmp_ins， 之前我看issue是有个用户反馈说，希望有一个直接使用的dygraph.grad() 接口来替换 loss.backward 这种写法

Comment (sneaxiy) [reply paddle/fluid/imperative/partial_grad_engine.cc:778]: 这个tmp_ins是必须创建的。因为`dygraph.grad`不能影响任何前向var的梯度，`dygraph.grad`计算过程中所有梯度都必须是全新的变量，与前向的任何var都没有关系。

File: paddle/fluid/imperative/partial_grad_engine.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/imperative/partial_grad_engine.cc]: 已改成`std::vector`。

**Review by phlrain — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
